src/EFCore.PG/Query/Internal/NpgsqlQuerySqlGenerator.cs

```diff
@@ -206,6 +206,85 @@ protected override Expression VisitSqlBinary(SqlBinaryExpression binary)
         }
     }
 
+    protected override Expression VisitDelete(DeleteExpression deleteExpression)
```

Comment:
@smitpatel here's the PG join logic for USING, let me know if you have any thoughts.

One thing that crossed my mind... This may be a bit of "translation" logic embedded in the query SQL generation, e.g. moving the join condition out to the predicate. We could consider having a provider-specific PostgresDeleteExpression, which corresponds exactly to the PG SQL (target table, optional USINGs, predicate); a post-processing visitor would then convert the relational expression (which is a thin wrapper over SelectExpression) into that thing. This would keep SQL generation very simple, only rendering into SQL. But I'm not sure it's worth it, things look OK as-is too. Let me know what you think.

Reply:
I agree that it is abuse of design and we shouldn't be creating SqlExpression here. Though alternative may not be easy. In a way the relational layer abuses the fact that either things are supported just like SELECT or not supported at all. This allows it to remember the whole SelectExpression as-is and which will run through null semantics and other optimizations just fine. If you decide to process to Using before hand then you would need to create a structure from scratch which is somewhat similar to SelectExpression and also take care of it in null semantics and other visitor. The cost may not be worth.

The few visitors I can think of which works due to current design
- Pruning (specifically for TPT/TPC & of projection) when they are being used in filtering condition.
- Null semantics
- FromSql in predicate
- value compensation when using bool value in predicate

Reply:
Makes sense, yeah - everyone would have to handle the new expression(s). I can see how that might not be worth it - you know best (maybe it's worth giving it a try and seeing how bad it really is). 

Another approach would be to keep SelectExpression all the way to the end, and convert to PostgresDeleteExpression only at the end, after all visitors. This would at least keep SQL generation clean, but may not be super valuable either.

Let me know. I'll probably not do something special for PG here - if in relational you prefer to keep as-is I'll do the same and keep the current "hack" in SQL generation.

Reply:
The last processing could work. In a sense `SearchConditionConvertingExpressionVisitor` in SqlServer works similarly that afterwards if you run any visitor which generates expression it has to be search condition/value corrected. So you can put as very last step in derived `RelationalParameterBasedSqlProcessor`.
It would also be helpful as it will modify and generate the sql tree only once per parameter cache entry.

Reply:
OK, I'll give that a try!

Reply:
Done, see https://github.com/npgsql/efcore.pg/pull/2449#discussion_r931983589